File: sdks/node/test/core/migrate.test.ts
// A deployment that gates DDL opens unmigrated and applies the schema itself.
//
// Until `migrate` has run there are no tables, so every query fails — that is
// the gate working, not a fault.

import { mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { afterEach, describe, expect, it } from "vitest";
import { Queue } from "../../src/index";

let tempDir: string | undefined;

afterEach(() => {
  if (tempDir) {
    // Best effort: the native SQLite handle is released on GC, which JS cannot
    // force, and Windows refuses to unlink a file that still has one open.
    try {
      rmSync(tempDir, { recursive: true, force: true, maxRetries: 5, retryDelay: 50 });
    } catch {
      // the OS reclaims it
    }
    tempDir = undefined;
  }
});

function dbPath(): string {
  tempDir ??= mkdtempSync(join(tmpdir(), "taskito-migrate-"));
  return join(tempDir, "q.db");
}

describe("explicit migrate", () => {
  it("applies the schema a gated open withheld", async () => {
    const queue = new Queue({ dbPath: dbPath(), autoMigrate: false });

    await expect(queue.stats()).rejects.toThrow(/no such table/);

    const report = queue.migrate();
    expect(report.applied.length).toBeGreaterThan(0);
    expect(report.workflowApplied.length).toBeGreaterThan(0);
    expect(report.schemaless).toBe(false);
    await queue.stats();

    const again = queue.migrate();
    expect(again.applied).toEqual([]);
    expect(again.workflowApplied).toEqual([]);
    expect(again.archivedJobs).toBe(0);
  });

  it("leaves only the workflow tables for an auto-migrated queue", () => {
    // Opening applies the core schema; workflow tables are built on first
    // workflow use, so an explicit migrate is what brings them forward.
    const queue = new Queue({ dbPath: dbPath() });

    const report = queue.migrate();
    expect(report.applied).toEqual([]);
    expect(report.workflowApplied.length).toBeGreaterThan(0);

    expect(queue.migrate().workflowApplied).toEqual([]);
  });
});

File: sdks/java/src/test/java/org/byteveda/taskito/MigrateTest.java
package org.byteveda.taskito;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.fasterxml.jackson.databind.ObjectMapper;
import java.nio.file.Path;
import java.util.Map;
import org.byteveda.taskito.internal.JniQueueBackend;
import org.byteveda.taskito.model.MigrationReport;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

/**
 * A deployment that gates DDL opens unmigrated and applies the schema itself.
 *
 * <p>Until migrate has run there are no tables, so every query fails — that is
 * the gate working, not a fault.
 */
class MigrateTest {

    private static String optionsFor(Path dir, boolean autoMigrate) throws Exception {
        return new ObjectMapper()
                .writeValueAsString(
                        Map.of("backend", "sqlite", "dsn", dir.resolve("t.db").toString(), "autoMigrate", autoMigrate));
    }

    @Test
    void anUnmigratedQueueAppliesItsOwnSchema(@TempDir Path dir) throws Exception {
        try (Taskito queue = Taskito.builder().open(JniQueueBackend.open(optionsFor(dir, false)))) {
            assertThrows(Exception.class, queue::stats, "no tables exist yet");

            MigrationReport report = queue.migrate();
            assertFalse(report.applied.isEmpty(), "the first migrate applies the whole history");
            assertFalse(report.schemaless);
            queue.stats();

            assertTrue(queue.migrate().isEmpty(), "a current database reports no work");
        }
    }

    @Test
    void anAutoMigratedQueueHasOnlyItsWorkflowTablesLeft(@TempDir Path dir) throws Exception {
        try (Taskito queue = Taskito.builder().open(JniQueueBackend.open(optionsFor(dir, true)))) {
            // Opening applies the core schema; the workflow tables are built on
            // first workflow use, so an explicit migrate is what brings them
            // forward for a deployment that wants no DDL at runtime.
            MigrationReport report = queue.migrate();
            assertTrue(report.applied.isEmpty(), "the core schema was applied at open");
            assertFalse(report.workflowApplied.isEmpty(), "workflow tables were still pending");

            assertTrue(queue.migrate().isEmpty(), "a second run has nothing left");
        }
    }
}
